### PR TITLE
Support loading dependencies from /lib in Function App root

### PR DESCRIPTION
**Pull request by kevinhillinger**
Related to: #48,  #46 

Adding support for the worker to take all dependencies and add them to the system class loader, making all dependencies available to child class loaders.

### CODE REVIEW

**Comment by kevinhillinger** [comment]
@JunyiYi Here are the results of the tests.

[pr50-smoke-startup-output.txt](https://github.com/Azure/azure-functions-java-worker/files/1712499/pr50-smoke-startup-output.txt)

[LocalTest.pr50.latest.postman_test_run.json.txt](https://github.com/Azure/azure-functions-java-worker/files/1712497/LocalTest.pr50.latest.postman_test_run.json.txt)
Validating with smoke tests. I used the 1.0.0-beta-2 of azure-functions-java-core.

**Comment by JunyiYi** [comment]
No regression. Tests are fine.

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/JavaWorkerClient.java [hunk 1]
```diff
@@ -12,24 +12,29 @@
 
 import com.microsoft.azure.webjobs.script.broker.*;
 import com.microsoft.azure.webjobs.script.handler.*;
+import com.microsoft.azure.webjobs.script.reflect.ClassLoaderProvider;
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/JavaWorkerClient.java]: Please replace these individual imports with wildcard "*".

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/JavaWorkerClient.java [hunk 2]
```diff
@@ -109,4 +114,5 @@ private synchronized void send(String requestId, MessageHandler<?, ?> marshaller
     private final ManagedChannel channel;
     private final AtomicReference<StreamingMessagePeer> peer;
     private final Map<StreamingMessage.ContentCase, Supplier<MessageHandler<?, ?>>> handlerSuppliers;
+    private DefaultClassLoaderProvider classPathProvider;
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/JavaWorkerClient.java]: If this is not supposed to be changed later, make it `final`.

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/broker/JavaFunctionBroker.java [hunk 3]
```diff
@@ -45,6 +51,20 @@ public void loadMethod(String id, String funcname, String jarPath, String method
...
     }
+    
+    private void addSearchPathsToClassLoader(FunctionMethodDescriptor function) throws IOException {
+    		URL jarUrl = new File(function.getJarPath()).toURI().toURL();
+    		classLoaderProvider.addUrl(jarUrl);
+    		function.getLibDirectory().ifPresent(d -> registerWithClassLoaderProvider(d));
+    }
+    
+    private void registerWithClassLoaderProvider(File libDirectory) {
+    		try {
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/broker/JavaFunctionBroker.java]: Indentation issue: mixed "tab" and "space".

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/broker/JavaFunctionBroker.java [hunk 4]
```diff
@@ -13,17 +16,20 @@
...
 
-    public void loadMethod(String id, String funcname, String jarPath, String methodName, Map<String, BindingInfo> bindings)
-            throws ClassNotFoundException, MalformedURLException, NoSuchMethodException {
-        if (id == null || jarPath == null || methodName == null) {
-            throw new NullPointerException("id, jarPath, methodName should not be null");
-        }
-        JavaMethodExecutor executor = new JavaMethodExecutor(funcname, jarPath, methodName, bindings);
-        this.methods.put(id, new ImmutablePair<>(funcname, executor));
+    public void loadMethod(FunctionMethodDescriptor descriptor, Map<String, BindingInfo> bindings) 
+    		throws ClassNotFoundException, NoSuchMethodException, IOException
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/broker/JavaFunctionBroker.java]: Prefer "Space" indentation.

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/broker/JavaFunctionBroker.java [hunk 5]
```diff
@@ -45,6 +51,20 @@ public void loadMethod(String id, String funcname, String jarPath, String method
     public Optional<String> getMethodName(String id) {
         return Optional.ofNullable(this.methods.get(id)).map(entry -> entry.left);
     }
+    
+    private void addSearchPathsToClassLoader(FunctionMethodDescriptor function) throws IOException {
+    		URL jarUrl = new File(function.getJarPath()).toURI().toURL();
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/broker/JavaFunctionBroker.java]: Prefer "Space" Indentation.

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/broker/JavaMethodExecutor.java [hunk 6]
```diff
@@ -2,55 +2,46 @@
...
-        String jarPath = StringUtils.trim(jar);
-        if (StringUtils.isBlank(jarPath)) {
-            throw new IllegalArgumentException("\"" + jar + "\" is not a qualified JAR file name");
-        }
-        URL jarUrl = Paths.get(jarPath).toUri().toURL();
-        URLClassLoader jarLoader = new URLClassLoader(new URL[] { jarUrl });
+    JavaMethodExecutor(FunctionMethodDescriptor descriptor, Map<String, BindingInfo> bindingInfos, ClassLoaderProvider classLoaderProvider)
+            throws MalformedURLException, ClassNotFoundException, NoSuchMethodException 
+    {
+    		descriptor.validateMethodInfo();
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/broker/JavaMethodExecutor.java]: Prefer "Space" Indentation.

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/description/FunctionMethodDescriptor.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
+package com.microsoft.azure.webjobs.script.description;
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/description/FunctionMethodDescriptor.java]: Prefer "Space" Indentation.

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/reflect/DefaultClassLoaderProvider.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            method.invoke(sysloader, new Object[] { url });
+        }
+        catch (Throwable t)
+        {
+            throw new IOException("Error adding " + url + " to system classloader");
+        }
+        
+    }
+	
+	private static final String systemClassLoaderAddUrlMethodName = "addURL";
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/reflect/DefaultClassLoaderProvider.java]: Prefer "ALL_UPPER_CASE_LETTER" name for constants.

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/reflect/ClassLoaderProvider.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
+package com.microsoft.azure.webjobs.script.reflect;
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/reflect/ClassLoaderProvider.java]: Prefer "Space" Indentation.

File: azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/reflect/DefaultClassLoaderProvider.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
+package com.microsoft.azure.webjobs.script.reflect;
```
Comment (JunyiYi) [inline azure-functions-java-worker/src/main/java/com/microsoft/azure/webjobs/script/reflect/DefaultClassLoaderProvider.java]: Prefer "Space" Indentation.

**Comment by kevinhillinger** [comment]
All tab indentations corrected

**Comment by xscript** [comment]
Add @jdneo for awareness.

**Pull request merged**